### PR TITLE
add parallel scan repeated matmul tensor utility

### PR DESCRIPTION
**Pull request by martinjankowiak**

### CODE REVIEW

**Comment by fritzo** [comment]
This is not parallel scan, it is merely optimal binary reduction. Can you instead implement standard generic optimal reduction, something like this or find it on some job application site on the web:
```
def reduce_power(op, value, power):
    assert isinstance(power, int) and power >= 0
    if power == 0:
        raise NotImpleementedError
    bits = []
    while power:
        if power & 1:
            bits.append(power)
        power <<= 1
    result = value
    for bit in reversed(bits[:-1]):
        result = op(result, result)
        if bit:
            result = op(result, value)
    return result
```

**Comment by martinjankowiak** [comment]
i've refactored this for simplicity. please lmk if the name of the method or something else can be improved

File: pyro/ops/tensor_utils.py [hunk 1]
```diff
@@ -61,3 +61,27 @@ def convolve(signal, kernel, mode='full'):
 
     start_idx = (padded_size - truncate) // 2
     return result[..., start_idx: start_idx + truncate]
+
+
+def parallel_scan_repeated_matmul(M, n):
```
Comment (fritzo) [inline pyro/ops/tensor_utils.py]: Can you rename this to just `repeated_matmul`? the "parallel scan" part is an implementation detail.

File: pyro/ops/tensor_utils.py [hunk 2]
```diff
@@ -61,3 +61,27 @@ def convolve(signal, kernel, mode='full'):
...
+
+    doubling_rounds = 0 if n <= 2 else math.ceil(math.log(n, 2)) - 1
+
+    result = torch.stack([M, torch.matmul(M, M)])
+
+    for i in range(doubling_rounds):
+        doubled = torch.matmul(result[-1, ...].unsqueeze(0), result)
+        result = torch.stack([result, doubled]).reshape(-1, *result.shape[1:])
+
+    return result[0:n, ...]
```
Comment (fritzo) [inline pyro/ops/tensor_utils.py]: nit: this just means `result[:n]`

File: pyro/ops/tensor_utils.py [hunk 3]
```diff
@@ -61,3 +61,27 @@ def convolve(signal, kernel, mode='full'):
...
+    """
+    assert M.size(-1) == M.size(-2), "Input tensors must satisfy M.size(-1) == M.size(-2)."
+    assert n > 0, "argument n to parallel_scan_repeated_matmul must be 1 or larger"
+
+    doubling_rounds = 0 if n <= 2 else math.ceil(math.log(n, 2)) - 1
+
+    result = torch.stack([M, torch.matmul(M, M)])
+
+    for i in range(doubling_rounds):
+        doubled = torch.matmul(result[-1, ...].unsqueeze(0), result)
```
Comment (fritzo) [inline pyro/ops/tensor_utils.py]: nit: just `torch.matmul(result[-1], result)` should work fine

**Comment by martinjankowiak** [comment]
addressed comments

**Pull request merged**